pkg/operator/aws_platform_service_location/controller.go

```diff
@@ -38,12 +49,92 @@ func NewController(operatorClient operatorv1helpers.OperatorClient, infraClient
-	currentInfra, err := c.infraClient.Infrastructures().Get(ctx, "cluster", metav1.GetOptions{})
+	obj, err := c.infraLister.Get("cluster")
 	if errors.IsNotFound(err) {
-		// log not found errors, but don't degrade?
-		klog.Warningf("%w", err)
+		syncCtx.Recorder().Warningf("AWSPlatformServiceLocationController", "Required infrastructures.%s/cluster not found", configv1.GroupName)
 		return nil
 	}
-	klog.Info(spew.Sprintf("current state: %s", currentInfra.Status))
+	if err != nil {
+		return err
+	}
+
+	currentInfra := obj.DeepCopy()
+	var platformName configv1.PlatformType
+	if pstatus := currentInfra.Status.PlatformStatus; pstatus != nil {
+		platformName = pstatus.Type
+	}
+	if len(platformName) == 0 {
+		syncCtx.Recorder().Warningf("AWSPlatformServiceLocationController", "Falling back to deprecated status.platform because infrastructures.%s/cluster status.platformStatus.type is empty", configv1.GroupName)
+		platformName = currentInfra.Status.Platform
+	}
+
+	if platformName != configv1.AWSPlatformType {
+		return nil // nothing to do here.
+	}
+
+	if !(currentInfra.Spec.PlatformSpec.Type == configv1.AWSPlatformType &&
+		currentInfra.Spec.PlatformSpec.AWS != nil &&
+		len(currentInfra.Spec.PlatformSpec.AWS.ServiceEndpoints) > 0) {
+		return nil // nothing to do here
+	}
+
+	services := currentInfra.Spec.PlatformSpec.AWS.ServiceEndpoints
+	if err := validateServiceEndpoints(services); err != nil {
+		syncCtx.Recorder().Warningf("AWSPlatformServiceLocationController", "Invalid spec.platformSpec.aws.serviceEndpoints provided for infrastructures.%s/cluster", configv1.GroupName)
+		return err
+	}
+
+	sort.Slice(services, func(i, j int) bool {
+		return services[i].Name < services[j].Name
+	})
+	if equality.Semantic.DeepEqual(currentInfra.Status.PlatformStatus.AWS.ServiceEndpoints, services) {
+		return nil // nothing to do now
+	}
+
+	currentInfra.Status.PlatformStatus.AWS.ServiceEndpoints = services
+	_, err = c.infraClient.UpdateStatus(ctx, currentInfra, metav1.UpdateOptions{})
 	return err
 }
+
+func validateServiceEndpoints(endpoints []configv1.AWSServiceEndpoint) error {
+	fldPath := field.NewPath("spec", "platformSpec", "aws", "serviceEndpoints")
+
+	allErrs := field.ErrorList{}
+	tracker := map[string]int{}
+	for idx, e := range endpoints {
+		fldp := fldPath.Index(idx)
+		if eidx, ok := tracker[e.Name]; ok {
+			allErrs = append(allErrs, field.Invalid(fldp.Child("name"), e.Name, fmt.Sprintf("duplicate service endpoint not allowed for %s, service endpoint already defined at %s", e.Name, fldPath.Index(eidx))))
+		} else {
+			tracker[e.Name] = idx
+		}
+
+		if err := validateServiceURL(e.URL); err != nil {
+			allErrs = append(allErrs, field.Invalid(fldp.Child("url"), e.URL, err.Error()))
+		}
+	}
+	return allErrs.ToAggregate()
+}
+
+var schemeRE = regexp.MustCompile("^([^:]+)://")
+
+func validateServiceURL(uri string) error {
+	endpoint := uri
+	if !schemeRE.MatchString(endpoint) {
```

Comment:
This should really be a parseable URL or it shouldn't.  The sometimes API seems "nicer" but it ultimately makes for complication.  Let's do a straight `url.Parse` and then have a list of schemes we allow.  I don't think empty should be allowed and I would require https until we have a reason not to.